test/libsolutil/FixedHash.cpp

```diff
@@ -0,0 +1,270 @@
+{
+
+static_assert(std::is_same<h160, FixedHash<20>>());
+static_assert(std::is_same<h256, FixedHash<32>>());
+
+BOOST_AUTO_TEST_SUITE(FixedHashTest)
+
+BOOST_AUTO_TEST_CASE(default_constructor)
+{
+	BOOST_CHECK_EQUAL(
+		FixedHash<1>{}.hex(),
+		"00"
+	);
+	BOOST_CHECK_EQUAL(
+		FixedHash<1>{}.size,
+		1
+	);
+	BOOST_CHECK_EQUAL(
+		FixedHash<8>{}.hex(),
+		"0000000000000000"
+	);
+	BOOST_CHECK_EQUAL(
+		FixedHash<8>{}.size,
+		8
+	);
+	BOOST_CHECK_EQUAL(
+		FixedHash<20>{}.hex(),
+		"0000000000000000000000000000000000000000"
+	);
+	BOOST_CHECK_EQUAL(
+		FixedHash<20>{}.size,
+		20
+	);
+	BOOST_CHECK_EQUAL(
+		FixedHash<32>{}.hex(),
+		"0000000000000000000000000000000000000000000000000000000000000000"
+	);
+	BOOST_CHECK_EQUAL(
+		FixedHash<32>{}.size,
+		32
+	);
+}
+
+BOOST_AUTO_TEST_CASE(bytes_constructor)
+{
+	FixedHash<8> a(bytes{});
+	BOOST_CHECK_EQUAL(a.size, 8);
+	BOOST_CHECK_EQUAL(a.hex(), "0000000000000000");
+
+	FixedHash<8> b(bytes{0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88});
+	BOOST_CHECK_EQUAL(b.size, 8);
+	BOOST_CHECK_EQUAL(b.hex(), "1122334455667788");
+
+	// TODO: short input, this should fail
+	FixedHash<8> c(bytes{0x11, 0x22, 0x33, 0x44, 0x55, 0x66});
+	BOOST_CHECK_EQUAL(c.size, 8);
+	BOOST_CHECK_EQUAL(c.hex(), "0000000000000000");
+
+	// TODO: oversized input, this should fial
+	FixedHash<8> d(bytes{0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99});
+	BOOST_CHECK_EQUAL(d.size, 8);
+	BOOST_CHECK_EQUAL(d.hex(), "0000000000000000");
+}
+
+// TODO: add FixedHash(bytesConstRef) constructor
+
+BOOST_AUTO_TEST_CASE(string_constructor_fromhex)
+{
+	// TODO: this tests the default settings onstructFromStringType::fromHex, ConstructFromHashType::FailIfDifferent
+	//       should test other options too
+
+	FixedHash<8> a("");
+	BOOST_CHECK_EQUAL(a.size, 8);
+	BOOST_CHECK_EQUAL(a.hex(), "0000000000000000");
+
+	FixedHash<8> b("1122334455667788");
+	BOOST_CHECK_EQUAL(b.size, 8);
+	BOOST_CHECK_EQUAL(b.hex(), "1122334455667788");
+
+	FixedHash<8> c("0x1122334455667788");
+	BOOST_CHECK_EQUAL(c.size, 8);
+	BOOST_CHECK_EQUAL(c.hex(), "1122334455667788");
+
+	// TODO: short input, this should fail
+	FixedHash<8> d("112233445566");
+	BOOST_CHECK_EQUAL(d.size, 8);
+	BOOST_CHECK_EQUAL(d.hex(), "0000000000000000");
+
+	// TODO: oversized input, this should fial
+	FixedHash<8> e("112233445566778899");
+	BOOST_CHECK_EQUAL(e.size, 8);
+	BOOST_CHECK_EQUAL(e.hex(), "0000000000000000");
+}
+
+BOOST_AUTO_TEST_CASE(string_constructor_frombytes)
+{
+
+	FixedHash<8> b("", FixedHash<8>::FromBinary);
+	BOOST_CHECK_EQUAL(b.size, 8);
+	BOOST_CHECK_EQUAL(b.hex(), "0000000000000000");
+
+	FixedHash<8> c("abcdefgh", FixedHash<8>::FromBinary);
+	BOOST_CHECK_EQUAL(c.size, 8);
+	BOOST_CHECK_EQUAL(c.hex(), "6162636465666768");
+
+	// TODO: short input, this should fail
+	FixedHash<8> d("abcdefg", FixedHash<8>::FromBinary);
+	BOOST_CHECK_EQUAL(d.size, 8);
+	BOOST_CHECK_EQUAL(d.hex(), "0000000000000000");
+
+	// TODO: oversized input, this should fail
```

Comment:
so currently all these `short` and `oversized` input just lead to zero?

Reply:
Yes. Well tested piece of code :)

Reply:
that's a weird behavior